esp32/modules/services.py

```diff
@@ -1,59 +1,250 @@
+        if rtcRequired and time.time() < 1482192000:
+            if not wifiFailed:
+                print("[SERVICES] RTC required, configuring...")
+                easyrtc.configure()
+            else:
+                print("[SERVICES] RTC required but not available. Skipping service.")
+                continue # Skip the app (because wifi failed and rtc not available)
+        
+        try:
+            srv.setup()
+        except BaseException as msg:
+            print("[SERVICES] Exception in service setup "+app+": ", msg)
+            continue
+        
+        if loopEnabled:
+            try:
+                loopCallbacks[srv.loop] = wifiInLoop
+            except:
+                print("[SERVICES] Loop requested but not defined in service "+app)
+            
+        if drawEnabled and drawCb:
+            try:
+                drawCallbacks.append(srv.draw)
+            except:
+                print("[SERVICES] Draw requested but not defined in service "+app)
+        
+        # Add the script to the global service list
+        services.append(srv)
+        
+    # Create loop timer
+    hasLoopTimer = False
+    global loopTimer
+    if len(loopCallbacks)>0:
+        print("[SERVICES] There are loop callbacks, starting loop timer!")
+        loopTimer = machine.Timer(0) #TODO: how to get this number?!?
+        loop_timer_callback(loopTimer)
+        hasLoopTimer = True
+        
+    # Create draw timer
+    hasDrawTimer = False
+    global drawTimer
+    if len(drawCallbacks)>0 and drawCb:
+        print("[SERVICES] There are draw callbacks, starting draw timer!")
+        drawTimer = machine.Timer(1) #TODO: how to get this number?!?
+        draw_timer_callback(drawTimer)
+        hasDrawTimer = True
+    return [hasLoopTimer, hasDrawTimer]
+            
+def loop_timer_callback(tmr):
+    global loopCallbacks
+    requestedInterval = 99999999
+    newLoopCallbacks = loopCallbacks
+    for cb in loopCallbacks:
+        if loopCallbacks[cb]:
+            print("[SERVICES] Loop needs wifi!")
+            if not easywifi.status():
+                if not easywifi.enable():
+                    print("[SERVICES] Wifi not available!")
+                    continue
+        else:
+            print("[SERVICES] Loop does not need wifi!")
+        rqi = 0
         try:
-            if (srv.loop(lcnt)):
-                noSleep = True
+            rqi = cb()
         except BaseException as msg:
-            print("[SERVICES] Service loop exception: ", msg)
-    return noSleep
+            print("[SERVICES] Exception in service loop: ", msg)
+            newLoopCallbacks.pop(cb)
+            continue
+        if rqi>0 and rqi<requestedInterval:
+            # Service wants to loop again in rqi seconds
+            requestedInterval = rqi
+        elif rqi<=0:
+            # Service doesn't want to loop again until next wakeup
+            newLoopCallbacks.pop(cb)
+    loopCallbacks = newLoopCallbacks
+    del(newLoopCallbacks)
```

Comment:
You can't change a dict while looping through it's items so I create a copy and change the copy. Then I replace the original with the copy and delete the copy.
The del doesn't do much but it frees a couple of bytes of ram.